1. Загрузите файл <code>Russian_ds.tsv</code>.
2. Создайте столбец с типом населенных пунктов.
3. Создайте столбец с названиями населенных пунктов, очищенный от типов.
   Созданные столбцы не должны содержать пропущенных значения, если невозможно установить тип населенного пункта, заполните пропуски. Допустимо заменить буквы ё на е для некоторых типов населенных пунктов.

Ожидаемый результат:

4. Получите список уникальных типов населенных пунктов, определите число уникальных типов населенных пунктов. Определите долю каждого типа населенного пункта. Для трех любых неизвестных для вас ранее типов населенных пунктов определите региональную принадлежность. 
5. Посчитайте сумму населения по типам населенных пунктов в абсолютных и относительных значениях.
6. Определите топ-10 самых популярных названий населенных пунктов. Определите число уникальных названий населенных пунктов.
7. Определите населенные пункты с самыми короткими названиями. Получите список населенных пунктов с длиной в два символа, найдите наиболее популярные названия из найденного списка. Определите региональную принадлежность. Повторите исследование для населенных пунктов с тремя символами.
8. Найдите населенный пункт с самым длинным названием из одного слова без пробелов и дефисов.
9. Определите число населенных пунктов на букву "Ы". На какую букву чаще всего начинаются названия населенных пунктов?
10. Определите сколько раз встречаются в названиях населенных пунктов прилагательные: большой, малый, красный, новый.
11. Найдите несколько примеров необычных названий населенных пунктов, связанных с одной из тематик (одной или нескольких, можно в разных). Допустимо предложить свою тематику.


Комментарии к коду, при необходимости, оформите используя #.<br>
Добавляйте комментарии в проект, оформите проект соблюдая структуру заданий.
<br> Разделы 4-11 сопроводите выводами.<br>
Нумерацию заданий, комментарии к заданиям и выводы по проекту оформите, используя текстовые ячейки.


Задание 1

Загрузите файл Russian_ds.tsv.

In [2]:
import pandas as pd
df = pd.read_csv('Russian_ds.tsv', '\t') #Т.к. файл формата tsv, то в качестве разделителя нужно указать '\t'

C:\Users\vkhom\AppData\Local\Temp\ipykernel_7744\1362889608.py:2: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument 'filepath_or_buffer' will be keyword-only.
  df = pd.read_csv('Russian_ds.tsv', '\t') #Т.к. файл формата tsv, то в качестве разделителя нужно указать '\t'


Задания 2, 3

Создайте столбец с типом населенных пунктов.

In [3]:
df.insert(1, 'type', None) #Добавление нового столбца. Нужны 3 параметра: индекс, имя столбца и значение
df.insert(2, 'only_name', None)

In [4]:
df['name'] = df['name'].str.replace('ё', 'е')
df['type'] = df['name'].str.extract('(поселок|казарма|разъезд|аул|кордон|железнодорожный остановочный пункт|хутор|деревня|село|населенный пункт|железнодорожная будка|железнодорожная станция|остановочная платформа|місто|станица|станция|железнодорожный разъезд|город|селище міського типу|местечко|территория|селище|починок|участок|слобода|просёлок|заимка|насёленный пункт|слободка|погост|железнодорожная площадка|территория|селище|остановочный пункт|высел|железнодорожный блокпост|заимка|улус|железнодорожная платформа|аал|блок-пост|обгонный пункт|железнодорожный путевой пост|арбан|кутан|поселение|полустанок|экопоселение|молочнотоварная ферма совхоза|остров|мыс|отделение)', expand=False)
df.type.fillna('другое', inplace=True)

Создайте столбец с названиями населенных пунктов, очищенный от типов.

In [5]:
df['only_name'] = df['name'].str.extract(r'(имени\D*|[А-Я]\D*|[Ё]\D*|№[0-9]\w*\D*|казарма\D*|[0-9]\w*\D*|джепель|Євпаторія)')

In [6]:
df.isna().sum()

name                0
type                0
only_name           0
other_name     154547
region_name         0
district            0
lon                 0
lat                 0
population          0
approx              0
dtype: int64

In [7]:
df.head()

name     type    only_name other_name            region_name  \
0   город Михайловка    город   Михайловка        NaN  Волгоградская область   
1  поселок Сибирский  поселок    Сибирский        NaN         Алтайский край   
2   хутор Никольский    хутор   Никольский        NaN   Белгородская область   
3      село Драгунка     село     Драгунка        NaN   Белгородская область   
4   село Нижние Пены     село  Нижние Пены        NaN   Белгородская область   

                            district        lon        lat  population  approx  
0   городской округ Город Михайловка  43.239730  50.070850     58407.0     0.0  
1  городской округ Посёлок Сибирский  83.839377  53.557357     11807.0     0.0  
2    Ракитянский муниципальный район  35.902320  51.005698        20.0     0.0  
3      Ивнянский муниципальный район  35.951177  51.022732       409.0     0.0  
4    Ракитянский муниципальный район  35.950410  51.005147      1000.0     0.0

Задание 4

Получите список уникальных типов населенных пунктов, определите число уникальных типов населенных пунктов. Определите долю каждого типа населенного пункта.

In [8]:
unique_types = pd.DataFrame({'quantity': df['type'].value_counts(),
                             'percentage': df['type'].value_counts(normalize=True).mul(100).round(2).astype('str') + '%'})
unique_types

quantity percentage
деревня                                91762     57.29%
село                                   30030     18.75%
поселок                                28165     17.58%
хутор                                   5734      3.58%
город                                   1170      0.73%
населенный пункт                        1070      0.67%
станица                                  444      0.28%
железнодорожная станция                  242      0.15%
станция                                  236      0.15%
аул                                      206      0.13%
улус                                     186      0.12%
починок                                  142      0.09%
разъезд                                  130      0.08%
слобода                                   82      0.05%
местечко                                  77      0.05%
железнодорожный разъезд                   76      0.05%
заимка                                    69      0.04%
кутан                                     64      0.04%
высел                                     51      0.03%
аал                                       46      0.03%
участок                                   46      0.03%
казарма                                   23      0.01%
другое                                    21      0.01%
кордон                                    21      0.01%
железнодорожная будка                     18      0.01%
арбан                                     12      0.01%
железнодорожный остановочный пункт        11      0.01%
остановочная платформа                     8       0.0%
слободка                                   6       0.0%
поселение                                  4       0.0%
блок-пост                                  3       0.0%
железнодорожный блокпост                   3       0.0%
селище міського типу                       3       0.0%
экопоселение                               3       0.0%
остановочный пункт                         2       0.0%
селище                                     2       0.0%
погост                                     2       0.0%
остров                                     2       0.0%
молочнотоварная ферма совхоза              2       0.0%
місто                                      1       0.0%
полустанок                                 1       0.0%
мыс                                        1       0.0%
железнодорожная платформа                  1       0.0%
отделение                                  1       0.0%
железнодорожная площадка                   1       0.0%
железнодорожный путевой пост               1       0.0%
обгонный пункт                             1       0.0%
территория                                 1       0.0%

Вывод: Чаще всего встречается тип "деревня" - 91762 раза, который занимает 57.29% от всех типов населённых пунктов

Для трех любых неизвестных для вас ранее типов населенных пунктов определите региональную принадлежность.

In [9]:
unusual_types = pd.DataFrame({'кутан': df.loc[df['type'] == 'кутан']['region_name'].unique(),
                              'аал': df.loc[df['type'] == 'аал']['region_name'].unique(),
                              'арбан': df.loc[df['type'] == 'арбан']['region_name'].unique()})
unusual_types

кутан                 аал            арбан
0  Республика Дагестан  Республика Хакасия  Республика Тыва

Вывод: кутан встречается в Республике Дагестан, аал - в Республике Хакасия, арбан - в Республике Тыва

Задание 5

Посчитайте сумму населения по типам населенных пунктов в абсолютных и относительных значениях.

In [10]:
df.groupby(['type'])['population'].sum().sort_values(ascending=False).reset_index() #Сумма в абсолютных значениях

type   population
0                                город  102902660.0
1                                 село   21788361.0
2                              поселок   16143447.0
3                              деревня    6557954.0
4                              станица    2154630.0
5                                хутор    1292738.0
6                                  аул     215741.0
7                                місто     107650.0
8                              слобода      93278.0
9                                 улус      82549.0
10             железнодорожная станция      71850.0
11                             станция      49780.0
12                    населенный пункт      44749.0
13                селище міського типу      22730.0
14                                 аал      14410.0
15                            местечко      10071.0
16                               кутан       8639.0
17                             починок       5211.0
18                             разъезд       4321.0
19                             участок       4075.0
20                               арбан       3034.0
21                               высел       2481.0
22                              заимка       2336.0
23                              другое       1880.0
24             железнодорожный разъезд       1651.0
25  железнодорожный остановочный пункт        498.0
26                            слободка        386.0
27                           отделение        300.0
28                              кордон        234.0
29               железнодорожная будка        162.0
30                             казарма        142.0
31                  остановочный пункт        118.0
32                              селище         86.0
33            железнодорожная площадка         83.0
34                           блок-пост         76.0
35                      обгонный пункт         60.0
36              остановочная платформа         47.0
37                           поселение         34.0
38                              остров         20.0
39           железнодорожная платформа         20.0
40                        экопоселение         14.0
41            железнодорожный блокпост         10.0
42                              погост         10.0
43        железнодорожный путевой пост         10.0
44                          территория          0.0
45                                 мыс          0.0
46       молочнотоварная ферма совхоза          0.0
47                          полустанок          0.0

In [11]:
#Сумма в относительных значениях (%)

df.insert(10, 'relative_value', None)
all_population = df['population']. sum ()
df['relative_value'] = df['population'] * 100 / all_population
df.groupby('type')['relative_value'].sum().sort_values(ascending=False).reset_index().round(2)

type  relative_value
0                                город           67.88
1                                 село           14.37
2                              поселок           10.65
3                              деревня            4.33
4                              станица            1.42
5                                хутор            0.85
6                                  аул            0.14
7                                місто            0.07
8                              слобода            0.06
9                                 улус            0.05
10             железнодорожная станция            0.05
11                             станция            0.03
12                    населенный пункт            0.03
13                селище міського типу            0.01
14                                 аал            0.01
15                            местечко            0.01
16                               кутан            0.01
17                             починок            0.00
18                             разъезд            0.00
19                             участок            0.00
20                               арбан            0.00
21                               высел            0.00
22                              заимка            0.00
23                              другое            0.00
24             железнодорожный разъезд            0.00
25  железнодорожный остановочный пункт            0.00
26                            слободка            0.00
27                           отделение            0.00
28                              кордон            0.00
29               железнодорожная будка            0.00
30                             казарма            0.00
31                  остановочный пункт            0.00
32                              селище            0.00
33            железнодорожная площадка            0.00
34                           блок-пост            0.00
35                      обгонный пункт            0.00
36              остановочная платформа            0.00
37                           поселение            0.00
38                              остров            0.00
39           железнодорожная платформа            0.00
40                        экопоселение            0.00
41            железнодорожный блокпост            0.00
42                              погост            0.00
43        железнодорожный путевой пост            0.00
44                          территория            0.00
45                                 мыс            0.00
46       молочнотоварная ферма совхоза            0.00
47                          полустанок            0.00

Вывод: 67,88% населения живут в городах. В абсолютном значении: 102 902 660 людей

Задание 6

Определите топ-10 самых популярных названий населенных пунктов.

In [12]:
#посчитаем количество повторяющихся значений only_name по всем столбцам  

df.groupby('only_name').count()

name  type  other_name  region_name  \
only_name                                                                
1 Мая                                 7     7           0            7   
1 совхоза Сортавальский               1     1           1            1   
1 шлюзе ББК                           1     1           1            1   
1-го Мая                              1     1           0            1   
1-го отделения совхоза Масловский     1     1           0            1   
...                                 ...   ...         ...          ...   
№17                                   1     1           0            1   
№2                                    2     2           0            2   
№3                                    1     1           0            1   
№3 племзавода Парижская Коммуна       1     1           0            1   
№49                                   1     1           0            1   

                                   district  lon  lat  population  approx  \
only_name                                                                   
1 Мая                                     7    7    7           7       7   
1 совхоза Сортавальский                   1    1    1           1       1   
1 шлюзе ББК                               1    1    1           1       1   
1-го Мая                                  1    1    1           1       1   
1-го отделения совхоза Масловский         1    1    1           1       1   
...                                     ...  ...  ...         ...     ...   
№17                                       1    1    1           1       1   
№2                                        2    2    2           2       2   
№3                                        1    1    1           1       1   
№3 племзавода Парижская Коммуна           1    1    1           1       1   
№49                                       1    1    1           1       1   

                                   relative_value  
only_name                                          
1 Мая                                           7  
1 совхоза Сортавальский                         1  
1 шлюзе ББК                                     1  
1-го Мая                                        1  
1-го отделения совхоза Масловский               1  
...                                           ...  
№17                                             1  
№2                                              2  
№3                                              1  
№3 племзавода Парижская Коммуна                 1  
№49                                             1  

[82097 rows x 10 columns]

In [13]:
#посчитаем количество повторяющихся значений only_name 

df.groupby('only_name')['only_name'].count()

only_name
1 Мая                                7
1 совхоза Сортавальский              1
1 шлюзе ББК                          1
1-го Мая                             1
1-го отделения совхоза Масловский    1
                                    ..
№17                                  1
№2                                   2
№3                                   1
№3 племзавода Парижская Коммуна      1
№49                                  1
Name: only_name, Length: 82097, dtype: int64

In [14]:
#упорядочим подсчитанные повторяющиеся значения по столбцу only_name в порядке возрастания

df.groupby('only_name')['only_name'].count().sort_values()

only_name
Матюхинский         1
Ошминское           1
Ошмянка             1
Ошняк               1
Ошняк-Качкалак      1
                 ... 
Никольское        240
Каменка           252
Михайловка        267
Ивановка          280
Александровка     331
Name: only_name, Length: 82097, dtype: int64

In [15]:
#выведем экран 10 самых часто повторяющихся названий населенных пунктов

df.groupby('only_name')['only_name'].count().sort_values().tail(10)

only_name
Заречье          198
Первомайский     202
Сосновка         216
Березовка        233
Николаевка       238
Никольское       240
Каменка          252
Михайловка       267
Ивановка         280
Александровка    331
Name: only_name, dtype: int64

In [16]:
#упорядочить в порядке убывания 10 самых часто повторяющихся названий населенных пунктов

df.groupby('only_name')['only_name'].count().sort_values(ascending = False).head(10)

only_name
Александровка    331
Ивановка         280
Михайловка       267
Каменка          252
Никольское       240
Николаевка       238
Березовка        233
Сосновка         216
Первомайский     202
Заречье          198
Name: only_name, dtype: int64

Вывод: Чаще всего встречается название "Александровка" - 331 раз. Дальше по убыванию идёт Ивановка, Михайловка, Каменка, Никольское, Николаевка, Березовка, Сосновка, Первомайский и Заречье

Определите число уникальных названий населенных пунктов.

In [17]:
#выведем на экран все уникальные названия населенных пунктов

df['only_name'].unique()

array(['Михайловка', 'Сибирский', 'Никольский', ..., 'Штормове', '1062',
       'Малое Сартово'], dtype=object)

In [18]:
#посчитаем все уникальные названия населенных пунктов

df['only_name'].nunique()

82097

Вывод: Итого в датафрейме 82097 уникальных названий населённых пунктов

Задание 7

Определите населенные пункты с самыми короткими названиями. Получите список населенных пунктов с длиной в два символа, найдите наиболее популярные названия из найденного списка.

In [20]:
df.loc[df['only_name'].str.len() == 2][['only_name']].value_counts().head(10)

# loc позволяет обратиться к элементам конкретного столбца (в данном случае к элементам only_name). 
#str.len() - выялвение количества символов в каждом элементе (создаем условие, что таких символов всег два в каждом элементе столбца)
#value.counts - считаем количество элементов, удовлетворяющим данному условию
#head(10) - выводим первые десять элементов, если хотим больше - ставим в скобочки желаемое число

only_name
Яр           11
Ям           11
Ик            4
№1            3
Ыб            3
№2            2
Ай            2
Уя            2
Ук            2
Ея            2
dtype: int64

Вывод: Самым популярным населенным пунктом оказался "Ям", на втором месте "Яр"

Определите региональную принадлежность для наиблее часто встречающихся названий.

In [25]:
#Наиболее часто встречаются у нас "Ям" и "Яр"

first_region = pd.DataFrame({'Яр': df.loc[df['only_name'] == 'Яр']['region_name'].unique()}) 
#df.loc[df['only_name'] == 'Яр' - для каждого значения столбца only_name найдем те строки, где содержится "Яр"
# далее для этих значений выводим соответствующие им уникальные данные региона (региональную принадлежность населенных пунктов)

second_region = pd.DataFrame({'Ям': df.loc[df['only_name'] == 'Ям']['region_name'].unique()})
#то же самое, но для 'Яра'

two_regions = second_region.join(first_region)
#объединяем первый датафрейм и второй
two_regions

Ям                    Яр
0  Белгородская область  Белгородская область

Ответ: Населеный пункт с названием "Ям" встречается в Волгоградской и Белгородской областях, а населенный пункт с названием "Ям" — только в Белгородской области

Повторите исследование для населенных пунктов с тремя символами.

In [98]:
# Получите список населенных пунктов с длиной в три символа, найдите наиболее популярные названия из найденного списка.

df.loc[df['only_name'].str.len() == 3][['only_name']].value_counts().head(10)

#аналогично

only_name
Бор          102
Дор           55
Мыс           29
Луч           16
Лом           13
Луг           11
Зуи           11
Низ           11
Рог            9
Жар            7
dtype: int64

Ответ: наиболее часто встречается "Бор", "Дор" и "Мыс"

In [99]:
#Определите региональную принадлежность для наиблее часто встречающихся названий.

#Наиболее часто встречаются у нас "Бор", 'Дор' и "Мыс"

#Составим датафрейм для Бора
three_region = pd.DataFrame({'Бор': df.loc[df['only_name'] == 'Бор']['region_name'].unique()}) 

#Составим датафрейм для Дора
three_region_2 = pd.DataFrame({'Дор': df.loc[df['only_name'] == 'Дор']['region_name'].unique()})

# Объединяем первый и второй датафреймы
three_region_2 = three_region_2.join(three_region)

# Составим датафрейм бля Мыса
three_region_3 = pd.DataFrame({'Мыс': df.loc[df['only_name'] == 'Мыс']['region_name'].unique()})

#Объединяем третий датафрейм с остальными
three_region_3 = three_region_3.join(three_region_2)
three_region_3

Мыс                   Дор                   Бор
0  Белгородская область  Белгородская область  Белгородская область

Вывод: "Мыс", "Дор" и "Бор" находятся в Белгородской области

Задание 8

Найдите населенный пункт с самым длинным названием из одного слова без пробелов и дефисов.

In [100]:
longest_word = max(sum(df['only_name'].str.split(r' |-'), []), key=len)
longest_word

'Верхненовокутлумбетьево'

In [101]:
words = []
for list in df['only_name'].str.split(r' |-'):
    words.extend(list)

longest_word2 = max(words, key=len)
longest_word2

'Верхненовокутлумбетьево'

Вывод: населённый пункт с самым длинным названием - Верхненовокутлумбетьево

Задание 9

Определите число населенных пунктов на букву "Ы".

In [102]:
df['only_name'].str[0].value_counts()['Ы']

22

Вывод: на букву "Ы" начинаются 22 населённых пункта

На какую букву чаще всего начинаются названия населенных пунктов?

In [103]:
df['only_name'].str[0].value_counts()

К    19539
С    15017
П    13064
Б    12542
М    11420
Н     9692
В     8473
Г     6365
Т     6344
А     6284
Л     5929
Д     5333
З     5139
Р     4728
О     4170
Ш     3537
Ч     3498
И     3411
У     2990
Х     2485
Е     2134
Я     1897
Ф     1800
Ж     1387
Ю      994
Ц      521
Щ      497
Э      275
1      181
2      159
и      134
к       45
3       44
Ё       38
Ы       22
4       21
5       19
8       14
№        9
9        9
Й        7
6        7
7        7
д        1
Є        1
Name: only_name, dtype: int64

Вывод: чаще всего названия населённых пунктов начинаются на букву "К"

Задание 10

Определите сколько раз встречаются в названиях населенных пунктов прилагательные: большой, малый, красный, новый.

In [104]:
#определим населенные пункты в названиях которых есть приагательное'большой'

df[df['only_name'].str.contains('большой|большая|большие', case=False)]

name     type  \
36403                    деревня Скрабы  деревня   
38081                  деревня Абрамиха  деревня   
44337   поселок Криушанские Выселки 1-е  поселок   
48521                   деревня Михеево  деревня   
52333                 деревня Анисимово  деревня   
...                                 ...      ...   
159529                       село Шюнда     село   
159821                  поселок Потешны  поселок   
159861                 поселок Алферово  поселок   
159936                  поселок Пыловка  поселок   
160118                  деревня Мешково  деревня   

                                   only_name  other_name  \
36403                                Большая         NaN   
38081                                Большая         NaN   
44337                                Большая         NaN   
48521                                Большая         NaN   
52333                                Большая         NaN   
...                                      ...         ...   
159529                Большая Константиновка  село Шунды   
159821              Большой Камыш (Галищева)         NaN   
159861             Большие Мордовские Пошаты         NaN   
159936            Новое Село (Большой Курят)         NaN   
160118  Остров Большой Олений (Русский Маяк)         NaN   

                 region_name                              district  \
36403       Тверская область   Западнодвинский муниципальный округ   
38081       Тверская область          Бежецкий муниципальный район   
44337    Воронежская область         Панинский муниципальный район   
48521    Ярославская область       Ярославский муниципальный район   
52333      Калужская область  Малоярославецкий муниципальный район   
...                      ...                                   ...   
159529  Чеченская Республика    Итум-Калинский муниципальный район   
159821     Кировская область    Верхошижемский муниципальный район   
159861    Московская область                 городской округ Чехов   
159936     Иркутская область        Слюдянский муниципальный район   
160118        Омская область   Большереченский муниципальный район   

               lon        lat  population  approx  relative_value  
36403    31.984740  56.164985        10.0     0.0        0.000007  
38081    36.366324  57.707852        10.0     0.0        0.000007  
44337    39.957446  51.445617        20.0     0.0        0.000013  
48521    39.459628  57.603674         4.0     1.0        0.000003  
52333    36.526341  55.061135        40.0     1.0        0.000026  
...            ...        ...         ...     ...             ...  
159529   45.359933  42.734522         4.0     0.0        0.000003  
159821   49.221924  57.984542         0.0     0.0        0.000000  
159861   37.244590  55.201130         7.0     0.0        0.000005  
159936  104.018244  51.758032         3.0     0.0        0.000002  
160118   74.687496  56.408359         0.0     0.0        0.000000  

[1405 rows x 11 columns]

In [105]:
#посчитаем населенные пункты в названиях которых есть приагательное'большой'

df[df['only_name'].str.contains('большой|большая|большие', case=False)].count()['only_name']

1405

In [106]:
#определим населенные пункты в названиях которых есть приагательное'малый'

df[df['only_name'].str.contains('малый|малая|малые', case=False)]

name                   type  \
3521                                деревня Бобровня                деревня   
3761                                поселок Каменный                поселок   
6875                                 деревня Блиниха                деревня   
9141                                  деревня Змиево                деревня   
9208                               деревня Сороченка                деревня   
...                                              ...                    ...   
159962  населенный пункт База Отдыха Комбината Искож       населенный пункт   
159967             железнодорожная будка 14 километр  железнодорожная будка   
159988                          деревня Новые Гусины                деревня   
160014                             деревня Шалабаево                деревня   
160034                                 село Міжводне                   село   

                             only_name other_name           region_name  \
3521                             Малый        NaN      Брянская область   
3761                             Малая        NaN      Брянская область   
6875                             Малая        NaN    Ивановская область   
9141                             Малая        NaN     Калужская область   
9208                             Малая        NaN     Калужская область   
...                                ...        ...                   ...   
159962     Малые Подовечинские Выселки        NaN     Кировская область   
159967     Мартемьянцево (Малые Пруды)        NaN     Кировская область   
159988     Усть-Малые Чирки (Казаково)        NaN  Новгородская область   
160014    Старый Кумляк (Малые Житари)        NaN      Тверская область   
160034  Малая Емельянова Горка (Подол)        NaN       Республика Крым   

                                   district        lon        lat  population  \
3521         Дубровский муниципальный район  33.451562  53.514912        40.0   
3761          Суражский муниципальный район  32.288222  53.046157         0.0   
6875            Лухский муниципальный район  42.633198  57.074830        46.0   
9141          Юхновский муниципальный район  35.402746  54.628500         4.0   
9208          Мещовский муниципальный район  35.059913  54.355941        10.0   
...                                     ...        ...        ...         ...   
159962  Кирово-Чепецкий муниципальный район  49.737377  58.298757         5.0   
159967          городской округ Город Киров  49.553602  58.636202         0.0   
159988         Батецкий муниципальный район  30.376889  58.605204         0.0   
160014         Осташковский городской округ  32.734430  57.279955         8.0   
160034     Черноморский муниципальный район  32.840649  45.574375      1602.0   

        approx  relative_value  
3521       0.0        0.000026  
3761       1.0        0.000000  
6875       0.0        0.000030  
9141       1.0        0.000003  
9208       0.0        0.000007  
...        ...             ...  
159962     0.0        0.000003  
159967     0.0        0.000000  
159988     0.0        0.000000  
160014     0.0        0.000005  
160034     0.0        0.001057  

[1261 rows x 11 columns]

In [107]:
#посчитаем населенные пункты в названиях которых есть прилагательное'малый'

df[df['only_name'].str.contains('малый|малая|малые', case=False)].count()['only_name']

1261

In [108]:
#определим населенные пункты в названиях которых есть приагательное'красный'

df[df['only_name'].str.contains('красный|красная|красные', case=False)]

name                     type  \
28492               деревня Черносвитово                  деревня   
28572                  деревня Петрищево                  деревня   
28632                     деревня Щукино                  деревня   
28796                   деревня Марьинка                  деревня   
28879               поселок Оленьковский                  поселок   
...                                  ...                      ...   
159697                      поселок Золх                  поселок   
159706         деревня Воловские Выселки                  деревня   
160046  железнодорожная станция Трясенка  железнодорожная станция   
160109                  деревня Морозово                  деревня   
160169                   деревня Волчиха                  деревня   

                                            only_name other_name  \
28492                                         Красный        NaN   
28572                                         Красный        NaN   
28632                                         Красный        NaN   
28796                                         Красная        NaN   
28879                                         Красный        NaN   
...                                               ...        ...   
159697                        Красный Кут (Усольский)        NaN   
159706                        Красные Речки (Дурново)        NaN   
160046                 Красный Октябрь (Краснопеевка)        NaN   
160109            Коломцы (Красная Заполярная Звезда)        NaN   
160169  Самарского отделения совхоза Красная Башкирия        NaN   

                          region_name  \
28492                Тульская область   
28572                Тульская область   
28632                Тульская область   
28796                Тульская область   
28879                Тульская область   
...                               ...   
159697  Кемеровская область — Кузбасс   
159706               Липецкая область   
160046               Тверская область   
160109             Московская область   
160169            Вологодская область   

                                       district        lon        lat  \
28492           Ясногорский муниципальный район  37.715745  54.713053   
28572              Заокский муниципальный район  37.277141  54.656875   
28632   муниципальное образование Город Алексин  37.009852  54.468135   
28796           Ясногорский муниципальный район  37.900586  54.482797   
28879             Венёвский муниципальный район  38.101236  54.575670   
...                                         ...        ...        ...   
159697        Прокопьевский муниципальный округ  86.495894  53.875026   
159706            Воловский муниципальный район  37.933832  52.008754   
160046           Торжокский муниципальный район  34.743073  57.014927   
160109                Раменский городской округ  38.369532  55.393011   
160169          Вожегодский муниципальный район  40.200117  60.559761   

        population  approx  relative_value  
28492         20.0     0.0        0.000013  
28572          4.0     1.0        0.000003  
28632         27.0     0.0        0.000018  
28796         10.0     0.0        0.000007  
28879        400.0     1.0        0.000264  
...            ...     ...             ...  
159697        11.0     0.0        0.000007  
159706         0.0     0.0        0.000000  
160046         0.0     0.0        0.000000  
160109         0.0     0.0        0.000000  
160169         0.0     0.0        0.000000  

[1224 rows x 11 columns]

In [109]:
#посчитаем населенные пункты в названиях которых есть прилагательное'красный'

df[df['only_name'].str.contains('красный|красная|красные', case=False)].count()['only_name']

1224

In [110]:
#определим населенные пункты в названиях которых есть приагательное'новый'


df[df['only_name'].str.contains('новый|новая|новые', case=False)]

name  \
3528                                  деревня Чепеничи   
3733                          деревня Красная Знаменка   
3740                            поселок Новые Далисичи   
3741                            поселок Новопокровский   
3742                                   деревня Нарость   
...                                                ...   
159411                                деревня Бирюлево   
159455  остановочный пункт 106 километр Рожневский Бор   
159629                                   поселок Сойга   
159884                    железнодорожная станция Пады   
160074                                деревня Горшково   

                           type                         only_name other_name  \
3528                    деревня                             Новый        NaN   
3733                    деревня                             Новая        NaN   
3740                    поселок                             Новая        NaN   
3741                    поселок                             Новая        NaN   
3742                    деревня                             Новая        NaN   
...                         ...                               ...        ...   
159411                  деревня             Горушка Новая Деревня        NaN   
159455       остановочный пункт             Военсовхоз Новая Ляда        NaN   
159629                  поселок           Починок (Новая Капешня)        NaN   
159884  железнодорожная станция         Новая Курья (Новый Строй)        NaN   
160074                  деревня  Новая Отрада (Отрада Башкирская)        NaN   

                  region_name                             district        lon  \
3528         Брянская область       Дубровский муниципальный район  33.439706   
3733         Брянская область        Суражский муниципальный район  32.466996   
3740         Брянская область        Суражский муниципальный район  32.408876   
3741         Брянская область        Суражский муниципальный район  32.485029   
3742         Брянская область        Суражский муниципальный район  32.452952   
...                       ...                                  ...        ...   
159411    Вологодская область      Вологодский муниципальный район  39.848381   
159455  Новосибирская область      Тогучинский муниципальный район  84.115669   
159629  Архангельская область          Ленский муниципальный район  48.126580   
159884    Саратовская область      Балашовский муниципальный район  43.150275   
160074    Вологодская область  Великоустюгский муниципальный район  46.125547   

              lat  population  approx  relative_value  
3528    53.629594        20.0     0.0        0.000013  
3733    53.183933        10.0     0.0        0.000007  
3740    53.148865        30.0     0.0        0.000020  
3741    53.116292         4.0     1.0        0.000003  
3742    53.122308        90.0     0.0        0.000059  
...           ...         ...     ...             ...  
159411  59.123766         0.0     0.0        0.000000  
159455  55.248505         0.0     0.0        0.000000  
159629  61.655424       411.0     0.0        0.000271  
159884  51.747331        74.0     0.0        0.000049  
160074  60.756479         0.0     0.0        0.000000  

[1582 rows x 11 columns]

In [111]:
#посчитаем населенные пункты в названиях которых есть прилагательное'новый'

df[df['only_name'].str.contains('новый|новая|новые', case=False)].count()['only_name']

1582

Задание 11

Найдите несколько примеров необычных названий населенных пунктов, связанных с одной из тематик.

In [116]:
topic_foreign = pd.DataFrame({
    'Тема': df[df['name'].str.contains('місто|селище міського типу',case=False)]['name'],
    'Регион': df[df['name'].str.contains('місто|селище міського типу',case=False)]['region_name'],
    'Численность человек': df[df['name'].str.contains('місто|селище міського типу',case=False)]['population']})
topic_foreign

# создали датафрейм с помощью словаря, где ключи - это названия столбцов
# обращаемся в самом датафрейме и в столбце 'name' ищем совпадения по заданным ключеным словам на иностранном языке. Там, где найены совпадения, выводим столбцы 'name', 'region_name', 'population'

Тема           Регион  \
159990  селище міського типу Чорноморське  Республика Крым   
160003                    місто Євпаторія  Республика Крым   
160105     селище міського типу Роздольне  Республика Крым   
160144        селище міського типу Мирний  Республика Крым   

        Численность человек  
159990              11356.0  
160003             107650.0  
160105               7165.0  
160144               4209.0

In [113]:
topic_foreign_2 = topic_foreign.T
topic_foreign_2

#транспонируем таблицу

159990           160003  \
Тема                 селище міського типу Чорноморське  місто Євпаторія   
Регион                                 Республика Крым  Республика Крым   
Численность человек                            11356.0         107650.0   

                                             160105  \
Тема                 селище міського типу Роздольне   
Регион                              Республика Крым   
Численность человек                          7165.0   

                                          160144  
Тема                 селище міського типу Мирний  
Регион                           Республика Крым  
Численность человек                       4209.0

In [114]:
topic_kazarma = pd.DataFrame({
    'Тема': df[df['name'].str.contains('казарма',case=False)]['name'],
    'Регион': df[df['name'].str.contains('казарма',case=False)]['region_name'],
    'Численность человек': df[df['name'].str.contains('казарма',case=False)]['population']})
topic_kazarma

#аналогично

Тема               Регион  \
12188            поселок казарма 489 километр  Костромская область   
12190            поселок казарма 493 километр  Костромская область   
25491            поселок казарма 368 километр   Смоленская область   
59822                          казарма 142 км     Хабаровский край   
59894                          казарма 286 км     Хабаровский край   
...                                       ...                  ...   
159960  железнодорожная казарма 1083 километр    Кировская область   
159963  железнодорожная казарма 1002 километр    Кировская область   
159964   железнодорожная казарма 907 километр    Кировская область   
159966    железнодорожная казарма 61 километр    Кировская область   
159968    железнодорожная казарма 17 километр    Кировская область   

        Численность человек  
12188                   4.0  
12190                  20.0  
25491                   0.0  
59822                   4.0  
59894                  10.0  
...                     ...  
159960                  5.0  
159963                  2.0  
159964                  0.0  
159966                  3.0  
159968                  0.0  

[74 rows x 3 columns]

In [115]:
topic_kazarma_2 = topic_kazarma.T
topic_kazarma_2

12188   \
Тема                 поселок казарма 489 километр   
Регион                        Костромская область   
Численность человек                           4.0   

                                           12190   \
Тема                 поселок казарма 493 километр   
Регион                        Костромская область   
Численность человек                          20.0   

                                           25491             59822   \
Тема                 поселок казарма 368 километр    казарма 142 км   
Регион                         Смоленская область  Хабаровский край   
Численность человек                           0.0               4.0   

                               59894             59902             59903   \
Тема                   казарма 286 км    казарма 207 км    казарма 213 км   
Регион               Хабаровский край  Хабаровский край  Хабаровский край   
Численность человек              10.0              10.0               4.0   

                               59948                        60574   \
Тема                    казарма 37 км  поселок казарма 25 километр   
Регион               Хабаровский край              Приморский край   
Численность человек              10.0                         36.0   

                                      64039   ...  \
Тема                            село Казарма  ...   
Регион               Республика Башкортостан  ...   
Численность человек                    520.0  ...   

                                                   159714  \
Тема                 железнодорожная казарма 839 километр   
Регион                                  Кировская область   
Численность человек                                   0.0   

                                                      159833  \
Тема                 населенный пункт Казарма 192-й километр   
Регион                                         Пермский край   
Численность человек                                      0.0   

                                                       159839  \
Тема                 населенный пункт Казарма 1509-й километр   
Регион                                          Пермский край   
Численность человек                                       0.0   

                                                    159958  \
Тема                 железнодорожная казарма 1040 километр   
Регион                                   Кировская область   
Численность человек                                    6.0   

                                                    159959  \
Тема                 железнодорожная казарма 1045 километр   
Регион                                   Кировская область   
Численность человек                                    2.0   

                                                    159960  \
Тема                 железнодорожная казарма 1083 километр   
Регион                                   Кировская область   
Численность человек                                    5.0   

                                                    159963  \
Тема                 железнодорожная казарма 1002 километр   
Регион                                   Кировская область   
Численность человек                                    2.0   

                                                   159964  \
Тема                 железнодорожная казарма 907 километр   
Регион                                  Кировская область   
Численность человек                                   0.0   

                                                  159966  \
Тема                 железнодорожная казарма 61 километр   
Регион                                 Кировская область   
Численность человек                                  3.0   

                                                  159968  
Тема                 железнодорожная казарма 17 километр  
Регион                                 Кировская область  
Численность человек                                  0.0  

[3 rows x 74 columns]